phpBB/phpbb/mention/source/base_user.php

```diff
@@ -0,0 +1,143 @@
+<?php
+/**
+ *
+ * This file is part of the phpBB Forum Software package.
+ *
+ * @copyright (c) phpBB Limited <https://www.phpbb.com>
+ * @license       GNU General Public License, version 2 (GPL-2.0)
+ *
+ * For full copyright and license information, please see
+ * the docs/CREDITS.txt file.
+ *
+ */
+
+namespace phpbb\mention\source;
+
+abstract class base_user implements source_interface
+{
+	/** @var \phpbb\db\driver\driver_interface */
+	protected $db;
+
+	/** @var \phpbb\config\config */
+	protected $config;
+
+	/** @var \phpbb\user_loader */
+	protected $user_loader;
+
+	/** @var string */
+	protected $phpbb_root_path;
+
+	/** @var string */
+	protected $php_ext;
+
+	/**
+	 * Constructor
+	 */
+	public function __construct(\phpbb\db\driver\driver_interface $db, \phpbb\config\config $config, \phpbb\user_loader $user_loader, $phpbb_root_path, $phpEx)
+	{
+		$this->db = $db;
+		$this->config = $config;
+		$this->user_loader = $user_loader;
+		$this->phpbb_root_path = $phpbb_root_path;
+		$this->php_ext = $phpEx;
+
+		if (!function_exists('phpbb_get_user_rank'))
+		{
+			include($this->phpbb_root_path . 'includes/functions_display.' . $this->php_ext);
+		}
+	}
+
+	/**
+	 * Builds a query based on user input
+	 *
+	 * @param string $keyword  Search string
+	 * @param int    $topic_id Current topic ID
+	 * @return string Query ready for execution
+	 */
+	abstract protected function query($keyword, $topic_id);
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function get_priority($row)
+	{
+		// By default every result from the source increases the priority by a fixed value
+		return 1;
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function get(array &$names, $keyword, $topic_id)
+	{
+		$fetched_all = false;
+		$keyword = utf8_clean_string($keyword);
+
+		// Do not query all possible users (just a moderate amount), cache results for 5 minutes
```

Comment:
you do not limit the size of the batch yourself, it is up to the query itself. So this comment should be moved to the query() method imo

Reply:
This comment is no longer valid, all user IDs are queried and cached.
Fixed that.